commons/src/main/java/org/infinispan/commons/api/functional/FunctionalMap.java

```diff
@@ -0,0 +1,634 @@
+package org.infinispan.commons.api.functional;
+
+import org.infinispan.commons.api.functional.EntryView.ReadEntryView;
+import org.infinispan.commons.api.functional.EntryView.ReadWriteEntryView;
+import org.infinispan.commons.api.functional.EntryView.WriteEntryView;
+import org.infinispan.commons.api.functional.Listeners.ReadWriteListeners;
+import org.infinispan.commons.api.functional.Listeners.WriteListeners;
+import org.infinispan.commons.util.CloseableIterator;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Top level functional map interface offering common functionality for the
+ * read-only, read-write, and write-only operations that can be run against a
+ * functional map.
+ *
+ * DESIGN RATIONALES:
+ * <ul>
+ *    <li>Originally, I tried to come up with a single functional map interface
+ *    that would encompass read-only, read-write and write-only operations, but
+ *    it felt quite bloated. By separating each major type of operations to its
+ *    own interface, it becomes easier to figure out which are all the read-only
+ *    operations, which are write-only...etc, which also helps the user quickly
+ *    see what they can do without being obstructed by other operation types.
+ *    </li>
+ *    <li>The conscious decision to separate read-only, write-only and
+ *    read-write interfaces helps massively with type safety. So, if a user
+ *    gets a read-only map, it can't write to it by mistake since no such
+ *    APIs are exposed. The same happens with write-only maps, the user can
+ *    only write and cannot make the mistake of reading from the entry view
+ *    because read operations are not exposed. The previous design did not
+ *    have this level of type safety and hence it relied on runtime exceptions
+ *    to catch these wrong combinations.</li>
```

Comment:
Not at all!!! Our Cache and JSR-107 are here to stay for a while. FunctionalMap offers an alternative for those wanting more advanced use cases, or want to use more functional async APIs.